CONTRIBUTING.md

```diff
@@ -47,6 +47,92 @@ Using commit signoffs tags is mandatory for all commits; we also
 encourage that each commit is small and cohesive. See the next sections for details.
 
 
+### Programming style
+
+#### Organization
+
+All code that interacts with the operating system, such as file access, network
+operations, timers, process creation, etc, must happen through a C++
+interface. These interfaces must be organized in a
+`common/<MODULE>/<MODULE>.hpp` hierarchy. The platform implementation must
+reside in `common/<MODULE>/impl/<MODULE>.cpp` files. It is allowed to use more
+files in addition to the `<MODULE>` named files, if necessary.
+
+All the API must be namespaced inside the `<MODULE>` name. Avoid C macros if
+possible, since they can't be namespaced.
+
+#### Code style
+
+Our code style follows [the Google C++ Style
+Guide](https://google.github.io/styleguide/cppguide.html) with a few
+exceptions. To run automatic code style formatting on files you modify, use our
+[clang-format
+template](https://github.com/mendersoftware/mendertesting/blob/master/.clang-format):
+
+```bash
+curl -O https://github.com/mendersoftware/mendertesting/raw/master/.clang-format
+clang-format -i --style=file:.clang-format <MODIFIED_FILES>
+```
+
+Note that the template requires clang-format version 15 or higher.
+
+##### C++ standard
+
+All shared code must use C++ features from no later than the C++11
```

Comment:
So it is explicitly 11, not 14? Because 14 essentially only was a clean-up release, whereas 17 really brought new stuff.

Reply:
@kacf One thing that I discovered recently is that `make_unique` is C++14 only (ref: https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique ). This might be important because without the helper we'll need to implement constructors in our classes specifically for `unique_ptr<T>`.

Reply:
Hmm, do you have an example of where this is needed? As far as I can figure, there are two main reasons it is needed:

1. When exceptions are enabled (which they won't be), and more than one function is being evaluated in the statement (see [this thread](https://stackoverflow.com/questions/53870522/why-use-stdmake-unique-in-c17) for a good explanation). In C++17 and later, they clarified the evaluation order, and this reason is moot from then onwards.
2. When passing arrays to the `unique_ptr` constructor, which it doesn't support, and erroneously accepts. I agree it would be nice to have the better type safety that `make_unique` provides in this context. But I think raw arrays will probably be rare, because anywhere I see them I would encourage people to use another `std` structure instead, like `vector`. Main reason is that you can enable range checking for them, which you can't for arrays.

Reply:
Are the Boost alternatives nicer?

Reply:
It looks like the Boost alternative is there exactly to fill the gap between C++11 and C++14, but is otherwise the same. So this we could indeed use.

The only (slight) disadvantage is that for historical reasons, their `make_unique` returns a `std::unique_ptr`, whereas their `make_shared` returns `boost::shared_ptr`. IOW we should not mix those two.

If we do use it, I would propose making a special exception for it:

```
using boost::make_unique;
```

Which gives several advantages:

* It means we can use it with its plain name, instead of prefixing it with `boost::`. This will make it much easier to transition to `std::make_unique` later once we move beyond C++11.
* We won't accidentally use `boost::make_shared`.

If that is acceptable I will add this to the style guide.